src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/VirtualProjectXmlProvider.cs

```diff
@@ -0,0 +1,109 @@
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class VirtualProjectXmlProvider(DotnetCliHelper dotnetCliHelper, ILoggerFactory loggerFactory)
+{
+    private readonly ILogger<VirtualProjectXmlProvider> _logger = loggerFactory.CreateLogger<VirtualProjectXmlProvider>();
+
+    internal async Task<(string VirtualProjectXml, ImmutableArray<SimpleDiagnostic> Diagnostics)?> GetVirtualProjectContentAsync(string documentFilePath, CancellationToken cancellationToken)
+    {
+        var workingDirectory = Path.GetDirectoryName(documentFilePath);
+        var process = dotnetCliHelper.Run(["run-api"], workingDirectory, shouldLocalizeOutput: true, redirectStandardInput: true);
+
+        cancellationToken.Register(() =>
+        {
+            process?.Kill();
+        });
+
+        var input = new RunApiInput.GetProject() { EntryPointFileFullPath = documentFilePath };
+        var inputJson = JsonSerializer.Serialize(input, RunFileApiJsonSerializerContext.Default.RunApiInput);
+        await process.StandardInput.WriteAsync(inputJson);
+        process.StandardInput.Close();
+
+        // Debug severity is used for these because we think it will be common for the user environment to have too old of an SDK for the call to work.
+        // Rather than representing a hard error condition, it represents a condition where we need to gracefully downgrade the experience.
+        process.ErrorDataReceived += (sender, args) => _logger.LogDebug($"dotnet run-api: {args.Data}");
+        process.BeginErrorReadLine();
+
+        var responseJson = await process.StandardOutput.ReadLineAsync(cancellationToken);
+        await process.WaitForExitAsync(cancellationToken);
+
+        if (process.ExitCode != 0)
+        {
+            _logger.LogDebug($"dotnet run-api exited with exit code '{process.ExitCode}'.");
+            return null;
+        }
+
+        if (string.IsNullOrWhiteSpace(responseJson))
+        {
+            _logger.LogError($"dotnet run-api exited with exit code 0, but did not return any response.");
+            return null;
+        }
+
+        try
+        {
+            var response = JsonSerializer.Deserialize(responseJson, RunFileApiJsonSerializerContext.Default.RunApiOutput);
+            if (response is RunApiOutput.Error error)
+            {
+                _logger.LogError($"dotnet run-api version: {error.Version}. Latest known version: {RunApiOutput.LatestKnownVersion}");
+                _logger.LogError($"dotnet run-api returned error: '{error.Message}'");
+                return null;
+            }
+
+            if (response is RunApiOutput.Project project)
+            {
+                if (project.Version > RunApiOutput.LatestKnownVersion)
+                {
+                    _logger.LogWarning($"'dotnet run-api' version '{project.Version}' is newer than latest known version {RunApiOutput.LatestKnownVersion}");
+                }
```

Comment:
I don't know why we'd do this, since it means any bump to the version means we're now broken. I commented on the run-api PR that we need to rethink the versioning generally, but minimally I'd delete this code.

Reply:
Because right now if you were to bump the version, you'll just immediately give up and fail. Without this, it might still work.

I'd say the feedback here is similar to global.json: the number of times I've had to had to go fetch an SDK to build another repo is infinitely higher than the number of times global.json actually helped me.